File: src/main/java/net/vektah/codeglance/config/ConfigService.java
/*
 * Copyright © 2013, Adam Scarr
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package net.vektah.codeglance.config;

import com.intellij.openapi.components.*;
import com.intellij.util.xmlb.XmlSerializerUtil;
import net.vektah.codeglance.Observable;
import org.jetbrains.annotations.Nullable;


@State(
	name = "CodeGlance",
	storages = {
		@Storage(id="other", file = StoragePathMacros.APP_CONFIG + "/CodeGlance.xml")
	}
)
public class ConfigService extends Observable<ConfigChangeListener> implements PersistentStateComponent<Config> {
	private Config config = new Config();

	public ConfigService() {
		super(ConfigChangeListener.class);
	}

	@Nullable @Override public Config getState() {
		return config;
	}

	@Override public void loadState(Config config) {
		XmlSerializerUtil.copyBean(config, this.config);
	}
}

File: src/main/java/net/vektah/codeglance/config/ConfigService.kt
package net.vektah.codeglance.config

import com.intellij.openapi.components.*
import com.intellij.util.xmlb.XmlSerializerUtil
import java.lang.ref.WeakReference

@State(
        name = "CodeGlance3",
        storages = [
            Storage( "CodeGlance.xml")
        ]
)
class ConfigService : PersistentStateComponent<Config> {
    private val observers : MutableList<WeakReference<() -> Unit>> = arrayListOf()
    private val config = Config()

    override fun getState(): Config? = config
    public fun onChange(f :() -> Unit) = observers.add(WeakReference<() -> Unit>(f))

    public fun notifyChange() {
        val it = observers.listIterator()
        while(it.hasNext()) {
            val f = it.next().get()

            if (f == null) {
                it.remove()
            } else {
                f()
            }
        }
    }

    override fun loadState(config: Config) {
        XmlSerializerUtil.copyBean(config, this.config)
    }
}

